File: apps/skillnet-a2a/src/skillnet_client.py
"""Async HTTP client for the SkillNet external API."""

import uuid

import httpx

from src.config import settings


class SkillNetClient:
    """Thin wrapper over SkillNet /ext/v1/ endpoints."""

    def __init__(self) -> None:
        self._base = settings.API_URL.rstrip("/")
        self._headers = {"Authorization": f"Bearer {settings.SKILLNET_API_KEY}"}

    def _client(self) -> httpx.AsyncClient:
        return httpx.AsyncClient(
            base_url=self._base,
            headers=self._headers,
            timeout=30.0,
        )

    async def who_knows(self, skill: str, min_level: str | None = None) -> dict:
        async with self._client() as client:
            params: dict = {"skill": skill}
            if min_level:
                params["min_level"] = min_level
            resp = await client.get("/ext/v1/skills/who-knows", params=params)
            resp.raise_for_status()
            return resp.json()

    async def get_gap(self) -> dict:
        async with self._client() as client:
            resp = await client.get("/ext/v1/skills/gaps")
            resp.raise_for_status()
            return resp.json()

    async def verify_skill(
        self, user_id: str, skill_name: str, level: str, source: str = "manual"
    ) -> dict:
        async with self._client() as client:
            resp = await client.post(
                "/ext/v1/skills/verify",
                json={
                    "user_id": user_id,
                    "skill_name": skill_name,
                    "level": level,
                    "source": source,
                },
            )
            resp.raise_for_status()
            return resp.json()

    async def list_skills(self) -> dict:
        async with self._client() as client:
            resp = await client.get("/ext/v1/skills")
            resp.raise_for_status()
            return resp.json()

    async def get_user_skills(self, user_id: str) -> dict:
        async with self._client() as client:
            resp = await client.get(f"/ext/v1/skills/users/{user_id}/skills")
            resp.raise_for_status()
            return resp.json()

    async def create_course(
        self,
        title: str,
        *,
        document_id: str | None = None,
        intent_density: int = 3,
        enroll_user_id: str | None = None,
        generate_artifacts: list[str] | None = None,
    ) -> dict:
        """Create a course end to end in one call.

        The end-to-end flow (propose schema, generate knowledge packs with retry,
        review, validate, prewarm) runs server-side and can take a few minutes on a
        real provider, so this request uses a long timeout instead of the default 30s.
        """
        payload: dict = {"title": title, "intent_density": intent_density}
        if document_id:
            payload["document_id"] = document_id
        if enroll_user_id:
            payload["enroll_user_id"] = enroll_user_id
        if generate_artifacts:
            payload["generate_artifacts"] = generate_artifacts
        async with httpx.AsyncClient(
            base_url=self._base, headers=self._headers, timeout=600.0
        ) as client:
            resp = await client.post("/ext/v1/courses/full", json=payload)
            resp.raise_for_status()
            return resp.json()

File: packages/skillnet-mcp/src/skillnetClient.ts
/**
 * Thin HTTP client for `/ext/v1`. This is the ONLY way this package talks to
 * SkillNet: plain `fetch` calls against the FastAPI backend, carrying the caller's
 * API key as a Bearer token. No PostgreSQL driver, no direct database access, no
 * business logic duplicated from `apps/skillnet-api` — every rule (org scoping, level
 * validation, gap thresholds, ...) lives server-side and this client just relays it.
 */
import { toApiError } from "./errors.js";

export interface SkillCategory {
  id: string;
  name: string;
  skills: Array<{ id: string; name: string; description?: string | null }>;
}

export interface WhoKnowsResponse {
  skill: string;
  employees: Array<{
    user_id: string;
    full_name: string;
    level: string;
    source: string;
    last_assessed_at?: string | null;
  }>;
  [key: string]: unknown;
}

export interface GapReport {
  gaps: Array<{
    skill: { id: string; name: string; category?: string };
    total_users: number;
    users_at_level: number;
    coverage_ratio: number;
    gap_severity: string;
    users_below?: Array<{ id: string; full_name: string; current_level: string }>;
  }>;
  [key: string]: unknown;
}

export interface UserSkill {
  skill_id: string;
  skill_name: string;
  category?: string;
  level: string;
  source: string;
  last_assessed_at?: string | null;
}

export interface CreateCourseParams {
  title: string;
  document_id?: string;
  intent_density?: number;
  enroll_user_id?: string;
  generate_artifacts?: string[];
  artifact_node_limit?: number;
}

export interface CreateCourseResult {
  course_id: string;
  title: string;
  schema_status: string;
  schema_version: number;
  node_count: number;
  packs_ready: number;
  packs_all_ready: boolean;
  packs_summary: string;
  nodes: Array<{ node_id: string; title: string; status: string }>;
  reviewed: boolean;
  validated: boolean;
  enrolled_user_id: string | null;
  prewarm_spawned: boolean;
  artifacts: Array<{ artifact_id: string; node_id: string; kind: string; status: string }>;
  warnings: string[];
  [key: string]: unknown;
}

export interface SkillNetClientOptions {
  apiUrl: string;
  apiKey: string;
  requestTimeoutMs?: number;
  createCourseTimeoutMs?: number;
}

export class SkillNetClient {
  private readonly apiUrl: string;
  private readonly apiKey: string;
  private readonly requestTimeoutMs: number;
  private readonly createCourseTimeoutMs: number;

  constructor(options: SkillNetClientOptions) {
    this.apiUrl = options.apiUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.requestTimeoutMs = options.requestTimeoutMs ?? 30_000;
    this.createCourseTimeoutMs = options.createCourseTimeoutMs ?? 600_000;
  }

  async listSkills(params: { category?: string; search?: string } = {}): Promise<
    SkillCategory[]
  > {
    return this.get("/ext/v1/skills", params, this.requestTimeoutMs);
  }

  async whoKnows(params: {
    skill: string;
    min_level?: string;
    limit?: number;
  }): Promise<WhoKnowsResponse> {
    return this.get("/ext/v1/skills/who-knows", params, this.requestTimeoutMs);
  }

  async getGap(params: {
    skill?: string;
    min_level?: string;
    threshold?: number;
  } = {}): Promise<GapReport> {
    return this.get("/ext/v1/skills/gaps", params, this.requestTimeoutMs);
  }

  async getUserSkills(userId: string): Promise<UserSkill[]> {
    return this.get(
      `/ext/v1/users/${encodeURIComponent(userId)}/skills`,
      {},
      this.requestTimeoutMs
    );
  }

  async createCourse(params: CreateCourseParams): Promise<CreateCourseResult> {
    return this.post("/ext/v1/courses/full", params, this.createCourseTimeoutMs);
  }

  // -- internals ---------------------------------------------------------------------

  private async get<T>(
    path: string,
    query: Record<string, string | number | undefined>,
    timeoutMs: number
  ): Promise<T> {
    const url = new URL(this.apiUrl + path);
    for (const [key, value] of Object.entries(query)) {
      if (value !== undefined && value !== null) {
        url.searchParams.set(key, String(value));
      }
    }
    const response = await this.fetchWithTimeout(url, { method: "GET" }, timeoutMs);
    if (!response.ok) {
      throw await toApiError(response);
    }
    return (await response.json()) as T;
  }

  private async post<T>(path: string, body: unknown, timeoutMs: number): Promise<T> {
    const url = new URL(this.apiUrl + path);
    const response = await this.fetchWithTimeout(
      url,
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(body),
      },
      timeoutMs
    );
    if (!response.ok) {
      throw await toApiError(response);
    }
    return (await response.json()) as T;
  }

  private async fetchWithTimeout(
    url: URL,
    init: RequestInit,
    timeoutMs: number
  ): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
      return await fetch(url, {
        ...init,
        signal: controller.signal,
        headers: {
          ...(init.headers ?? {}),
          authorization: `Bearer ${this.apiKey}`,
        },
      });
    } finally {
      clearTimeout(timer);
    }
  }
}
